Source/WebCore/editing/cocoa/AttributedString.mm

```diff
@@ -85,79 +89,202 @@
 
-        auto identifierPair = styleInfo.tableBlockAndTableIDs[index];
-        if (!identifierPair)
-            continue;
+    [table setWidth:item.marginMinXEdge type:NSTextBlockAbsoluteValueType forLayer:NSTextBlockMargin edge:NSMinXEdge];
+    [table setWidth:item.marginMinYEdge type:NSTextBlockAbsoluteValueType forLayer:NSTextBlockMargin edge:NSMinYEdge];
+    [table setWidth:item.marginMaxXEdge type:NSTextBlockAbsoluteValueType forLayer:NSTextBlockMargin edge:NSMaxXEdge];
+    [table setWidth:item.marginMaxYEdge type:NSTextBlockAbsoluteValueType forLayer:NSTextBlockMargin edge:NSMaxYEdge];
 
-        auto [tableBlockID, tableID] = *identifierPair;
-        auto ensureTableBlockResult = tableBlocks.ensure(tableBlockID, [&] {
-            return static_cast<NSTextTableBlock *>(block);
-        });
+    if (item.backgroundColor)
+        [table setBackgroundColor:item.backgroundColor->get()];
 
-        auto tableBlock = ensureTableBlockResult.iterator->value;
-        if (!tableBlock.table)
-            continue;
+    if (item.borderMinXEdgeColor)
+        [table setBorderColor:item.borderMinXEdgeColor->get() forEdge:NSMinXEdge];
+
+    if (item.borderMinYEdgeColor)
+        [table setBorderColor:item.borderMinYEdgeColor->get() forEdge:NSMinYEdge];
+
+    if (item.borderMaxXEdgeColor)
+        [table setBorderColor:item.borderMaxXEdgeColor->get() forEdge:NSMaxXEdge];
+
+    if (item.borderMaxYEdgeColor)
+        [table setBorderColor:item.borderMaxYEdgeColor->get() forEdge:NSMaxYEdge];
+}
 
-        auto ensureTableResult = tables.ensure(tableID, [&] {
-            return tableBlock.table;
+inline static RetainPtr<NSParagraphStyle> reconstructStyle(const ParagraphStyle& style, IdentifierToTableMap& tables, IdentifierToTableBlockMap& tableBlocks, IdentifierToListMap& lists)
+{
+    for (const auto& item : style.textLists) {
+        lists.ensure(item.thisID, [&] {
+            RetainPtr list = adoptNS([[PlatformNSTextList alloc] initWithMarkerFormat:item.markerFormat options:0]);
+            [list setStartingItemNumber:item.startingItemNumber];
+            return list;
         });
+    }
 
-        auto table = ensureTableResult.iterator->value;
-        RetainPtr<NSTextTableBlock> replacementBlock;
-        if (!ensureTableBlockResult.isNewEntry)
-            replacementBlock = tableBlock;
-        else if (!ensureTableResult.isNewEntry) {
-            replacementBlock = adoptNS([[PlatformNSTextTableBlock alloc] initWithTable:table startingRow:tableBlock.startingRow rowSpan:tableBlock.rowSpan startingColumn:tableBlock.startingColumn columnSpan:tableBlock.columnSpan]);
-            [replacementBlock _takeValuesFromTextBlock:tableBlock];
-            tableBlocks.set(tableBlockID, replacementBlock.get());
-        }
+    for (const auto& item : style.textTables) {
+        tables.ensure(item.thisID, [&] {
+            RetainPtr table = adoptNS([(NSTextTable *)[PlatformNSTextTable alloc] init]);
```

Comment:
cast seems unnecessary.  Would this work?
RetainPtr table = adoptNS([PlatformNSTextTable new])